nsqd/topic.go

```diff
@@ -145,23 +145,29 @@ func (t *Topic) GetExistingChannel(channelName string) (*Channel, error) {
 
 // DeleteExistingChannel removes a channel from the topic only if it exists
 func (t *Topic) DeleteExistingChannel(channelName string) error {
-	t.Lock()
+	t.RLock()
 	channel, ok := t.channelMap[channelName]
 	if !ok {
-		t.Unlock()
+		t.RUnlock()
 		return errors.New("channel does not exist")
 	}
-	delete(t.channelMap, channelName)
-	// not defered so that we can continue while the channel async closes
-	numChannels := len(t.channelMap)
-	t.Unlock()
+	t.RUnlock()
 
 	t.nsqd.logf(LOG_INFO, "TOPIC(%s): deleting channel %s", t.name, channel.name)
 
 	// delete empties the channel before closing
 	// (so that we dont leave any messages around)
+	//
+	// we do this before removing the channel from map below (with no lock)
+	// so that any incoming subs will error and not create a new channel
```

Comment:
it seems like, for normal non-ephemeral channels, there are no checks if the channel is exiting, in either `Channel.AddClient()` or `ProtocolV2.SUB()`

Reply:
Fixed this in https://github.com/nsqio/nsq/pull/1314/commits/aadef464b04b173bb9af02051e836982ceec0369 PTAL